text/0000-optional-jquery.md

```diff
@@ -0,0 +1,248 @@
+apps
+
+## Detailed design
+
+### Remove internal jQuery usage
+
+As of writing this, there are [major efforts](https://github.com/emberjs/ember.js/issues/16058) underway to remove and 
+cleanup the Ember codebase and especially its tests from jQuery usage. Having a way to fully test Ember without jQuery 
+is a prerequisite to officially support jQuery being optional. When this is done, it will enable a "no jQuery" mode, 
+that will make it not use jQuery anymore, but only native DOM APIs.
+
+### Add an opt-out flag
+
+There should be a global flag that will toggle the optional jQuery integration (true by default). When this is disabled,
+it will make Ember CLI's build process *not* include jQuery into the `vendor.js` bundle, *and* it will explicitly put 
+Ember itself into its "no jQuery" mode.
+
+The flag itself will not be made a public API. Rather it will be handled by a privileged addon, that will allow to 
+disable the integration flag, thus to opt out from jQuery integration. This approach is in line with 
+[RFC 278](https://github.com/emberjs/rfcs/pull/278) and [RFC 280](https://github.com/emberjs/rfcs/pull/280), to allow 
+for some better implementation flexibility.
+
+### Introduce `@ember/jquery` package
+
+Currently Ember CLI itself is importing jQuery into the app's `vendor.js` file. To decouple it from this task, and
+to allow for some better flexibility in the future, the responsibility for importing jQuery is moved to a dedicated 
+`@ember/jquery` addon.
+
+To not create any breaking changes, Ember CLI will have to check the app's dependencies for the presence of this addon. 
+If it is not present, it will continue importing jQuery *unless* the jQuery integration flag is disabled.
+In the other case, it will stop importing jQuery at all, and delegate this responsibility to the addon.
+
+To nudge users to install `@ember/jquery` when they need jQuery, some warning/deprecation messages should be issued when
+the addon is *not* installed and the integration flag is either not specified or is set to true. Only in the case the 
+app is actively opting out of jQuery integration the addon is not needed.
+
+The addon itself has to make sure the Ember CLI version in use is at least the one that introduced the above mentioned
+logic, to prevent importing jQuery twice.
+
+### Assertions for jQuery based APIs
+
+Apart from testing (see below), Ember features some APIs that directly expose jQuery, which naturally cannot continue 
+to work without it. For these APIs some assertions have to be added when running in "no jQuery" mode (and not in 
+production), that provide some useful error messages for the developer:
+
+* `Ember.$()`
+  should throw an assertion stating that jQuery is not available.
+* `this.$()` in components
+  should throw an assertion stating that jQuery is not available and that `this.element` and native DOM APIs should be
+  used instead.
```

Comment:
Should these throw deprecation warnings before jquery is removed?

Reply:
No, I don't think so. Importantly, this RFC is _not_ suggesting that usage of jQuery specific features are deprecated.  It's providing for a way to opt-out of that feature set which is coherent.

There certainly may be an additional deprecation RFC down the road, but for now we are trying to ensure that either option (using jQuery integration features or _not_ including jQuery) are viable coherent options for Ember apps...

Reply:
Exactly. I had a second follow up RFC planned, that is more aggressive (and thus maybe controversial) about pushing away from jQuery, including deprecations as mentioned. But this would be *complimentary* to this one, so basically the second step (if adopted).